rllib/models/specs/specs_dict.py

```diff
@@ -127,4 +138,151 @@ def validate(
+            at least the elements of the spec, but can contain more entries.
+        output_exact_match: If True, the output data (should be a nested dict) must
+            match the spec exactly. Otherwise, the data is validated as long as it
+            contains at least the elements of the spec, but can contain more entries.
+
+    Returns:
+        A wrapped instance method. In case of `cache=True`, after the first invokation
+        of the decorated method, the intance will have `__checked_specs_cache__`
+        attribute that store which method has been invoked at least once. This is a
+        special attribute can be used for the cache itself. The wrapped class method
+        also has a special attribute `__checked_specs__` that marks the method as
+        decorated.
+    """
+
+    if not input_spec and not output_spec:
+        raise ValueError("At least one of input_spec or output_spec must be provided.")
+
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(self, input_data, **kwargs):
+
+            if cache and not hasattr(self, "__checked_specs_cache__"):
+                self.__checked_specs_cache__ = {}
+
+            def should_validate():
+                return not cache or func.__name__ not in self.__checked_specs_cache__
+
+            def validate(data, spec, exact_match, tag="data"):
+                is_mapping = isinstance(spec, ModelSpec)
+                is_tensor = isinstance(spec, TensorSpec)
+
+                if is_mapping:
+                    if not isinstance(data, Mapping):
+                        raise ValueError(
+                            f"{tag} must be a Mapping, got {type(data).__name__}"
+                        )
+                    data = NestedDict(data)
+
+                if should_validate():
+                    try:
+                        if is_mapping:
+                            spec.validate(data, exact_match=exact_match)
+                        elif is_tensor:
+                            spec.validate(data)
+                    except ValueError as e:
+                        raise ValueError(
+                            f"{tag} spec validation failed on "
+                            f"{self.__class__.__name__}.{func.__name__}, {e}."
+                        )
+
+                    if not (is_tensor or is_mapping):
+                        if not isinstance(data, spec):
+                            raise ValueError(
+                                f"Input spec validation failed on "
+                                f"{self.__class__.__name__}.{func.__name__}, "
+                                f"expected {spec.__name__}, got "
+                                f"{type(data).__name__}."
+                            )
+                return data
+
+            input_data_ = input_data
+            if input_spec:
+                input_spec_ = getattr(self, input_spec)()
```

Comment:
can input_spec_ just be a member variable on self?
or maybe we can simply assume that the specs will be provided by RLModule under some hardcoded key names.
so by default, folks can simply use the decorator without specifying any parameters.

Reply:
That will assume some hard-coded names on the base class and is not the intention of this decorator. The decorator let's the user choose their own spec names. You will see the use-case in the [RLModule PR](https://github.com/ray-project/ray/pull/29449).